File: android/src/main/kotlin/kr/yhs/flutter_kakao_maps/controller/KakaoMapControllerHandler.kt
package kr.yhs.flutter_kakao_maps.controller

import com.kakao.vectormap.GestureType
import com.kakao.vectormap.camera.CameraPosition
import com.kakao.vectormap.camera.CameraUpdate
import com.kakao.vectormap.camera.CameraAnimation
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodCall
import kr.yhs.flutter_kakao_maps.converter.CameraTypeConverter.asCameraAnimation
import kr.yhs.flutter_kakao_maps.converter.CameraTypeConverter.asCameraUpdate
import kr.yhs.flutter_kakao_maps.converter.PrimitiveTypeConverter.asMap
import kr.yhs.flutter_kakao_maps.converter.PrimitiveTypeConverter.asInt
import kr.yhs.flutter_kakao_maps.converter.PrimitiveTypeConverter.asBoolean
import kr.yhs.flutter_kakao_maps.model.KakaoMapEvent

interface KakaoMapControllerHandler {
    fun handle(call: MethodCall, result: MethodChannel.Result) = when (call.method) {
        "getCameraPosition" -> getCameraPosition(result::success)
        "moveCamera" -> {
            val arguments = call.arguments?.asMap<Any?>()
            val animation = arguments?.get("cameraAnimation")?.asCameraAnimation()
            val camera = arguments?.get("cameraUpdate")!!.asCameraUpdate()
            moveCamera(camera, animation, result::success)
        }
        "setGestureEnable" -> {
            val arguments = call.arguments!!.asMap<Any?>()
            val gestureType = GestureType.getEnum(arguments["gestureType"]!!.asInt())
            val enable = arguments["enable"]!!.asBoolean()
            setGestureEnable(gestureType, enable, result::success)
        }
        "setEventHandler" -> setEventHandler(call.arguments!!.asInt())
        else -> result.notImplemented()
    }

    fun getCameraPosition(onSuccess: (cameraPosition: Map<String, Any>) -> Unit)

    fun moveCamera(
        cameraUpdate: CameraUpdate,
        cameraAnimation: CameraAnimation?,
        onSuccess: (Any?) -> Unit
    )

    fun setGestureEnable(
        gestureType: GestureType,
        enable: Boolean,
        onSuccess: (Any?) -> Unit
    )

    fun setEventHandler(
        event: Int
    )
}

File: ios/Classes/controller/KakaoMapControllerHandler.swift
import Flutter
import KakaoMapsSDK

protocol KakaoMapControllerHandler {
    var kakaoMap: KakaoMap { get }

    func getCameraPosition(onSuccess: @escaping (_ cameraPosition: [String: Any]) -> Void)

    func moveCamera(
        cameraUpdate: CameraUpdate,
        cameraAnimation: CameraAnimationOptions?,
        onSuccess: (Any?) -> Void
    )

    func setEventHandler(event: UInt8)

    func setGestureEnable(gestureType: GestureType, enable: Bool, onSuccess: (Any?) -> Void)

    func getBuildingHeightScale(onSuccess: (Float) -> Void)

    func setBuildingHeightScale(scale: Float, onSuccess: (Any?) -> Void)

    func fromScreenPoint(point: CGPoint, onSuccess: ([String:Double]) -> Void)

    func toScreenPoint(position: MapPoint, onSuccess: ([String:Double]) -> Void)

    func clearCache(onSuccess: (Any?) -> Void)

    func clearDiskCache(onSuccess: (Any?) -> Void)

    func canPositionVisible(zoomLevel: Int, position: [MapPoint], onSuccess: (Bool) -> Void)

    func changeMapType(mapType: String, onSuccess: (Any?) -> Void)

    func overlayVisible(overlayType: String, visible: Bool, onSuccess: (Any?) -> Void)

    func defaultGUIvisible(type: DefaultGUIType, visible: Bool, onSuccess: (Any?) -> Void)

    func defaultGUIposition(type: DefaultGUIType, gravity: GuiAlignment, position: CGPoint, onSuccess: (Any?) -> Void)

    func scaleAutohide(autohide: Bool, onSuccess: (Any?) -> Void)

    func scaleAnimationTime(fadeIn: Int, fadeOut: Int, retention: Int, onSuccess: (Any?) -> Void)
}

extension KakaoMapControllerHandler {
    func handle(call: FlutterMethodCall, result: @escaping FlutterResult) {
        let arguments = call.arguments as? [String: Any]
        switch call.method {
        case "getCameraPosition": getCameraPosition(onSuccess: result)
        case "moveCamera":
            let cameraUpdate = asCameraUpdate(kakaoMap: kakaoMap, payload: asDict(arguments!["cameraUpdate"]!))
            let rawCameraAnimation = castSafty(arguments!["cameraAnimation"], caster: asDict)
            let cameraAnimation = rawCameraAnimation != nil ? CameraAnimationOptions(payload: rawCameraAnimation!) : nil
            moveCamera(cameraUpdate: cameraUpdate, cameraAnimation: cameraAnimation, onSuccess: result)
        case "setEventHandler": setEventHandler(event: (call.arguments! as! UInt8))
        case "setGestureEnable":
            setGestureEnable(
                gestureType: GestureType(rawValue: asInt(arguments!["gestureType"]!))!,
                enable: asBool(arguments!["enable"]!),
                onSuccess: result
            )
        case "getBuildingHeightScale": getBuildingHeightScale(onSuccess: result)
        case "setBuildingHeightScale": setBuildingHeightScale(scale: asFloat(arguments!["scale"]!), onSuccess: result)
        case "fromScreenPoint":
            let pointPayload = asDictTyped(call.arguments!, caster: asDouble)
            fromScreenPoint(point: CGPoint(payload: pointPayload), onSuccess: result)
        case "toScreenPoint": toScreenPoint(position: MapPoint(payload: arguments!), onSuccess: result)
        case "clearCache": clearCache(onSuccess: result)
        case "clearDiskCache": clearDiskCache(onSuccess: result)
        case "canPositionVisible": 
            let zoomLevel = asInt(arguments!["zoomLevel"]!)
            let position = asArray(arguments!["position"]!, caster: { MapPoint(payload: asDict($0)) })
            canPositionVisible(zoomLevel: zoomLevel, position: position, onSuccess: result)
        case "changeMapType": changeMapType(mapType: asString(arguments!["mapType"]!), onSuccess: result)
        case "overlayVisible": 
            overlayVisible(
                overlayType: asString(arguments!["overlayType"]!), 
                visible: asBool(arguments!["visible"]!), 
                onSuccess: result
            )
        default: result(FlutterMethodNotImplemented)
        }
    }
}
